### PR TITLE
1021/http outputs endpoint

### PR DESCRIPTION
**Pull request by libscott**

### CODE REVIEW

File: tests/web/test_transactions.py [hunk 1]
```diff
@@ -180,3 +181,41 @@ def test_post_invalid_transfer_transaction_returns_400(b, client, user_pk):
         InvalidSignature.__name__, 'Transaction signature is invalid.')
     assert res.status_code == expected_status_code
     assert res.json['message'] == expected_error_message
+
+
+def test_transactions_get_list_good(client):
+    from functools import partial
```
Comment (sbellem) [inline tests/web/test_transactions.py]: std lib imports can go at the top

File: tests/db/test_bigchain_api.py [hunk 2]
```diff
@@ -1156,3 +1157,36 @@ def test_get_spent_multiple_owners(self, b, user_sk, user_pk):
...
+    from bigchaindb.core import Bigchain
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
+        b = Bigchain()
+        res = b.get_owned_ids("abc")
+    gof.assert_called_once_with("abc", include_spent=False)
+    assert res == gof()
+
+
+def test_get_outputs_filtered_only_unspent():
+    from bigchaindb.common.transaction import TransactionLink as TL
```
Comment (TimDaub) [inline tests/db/test_bigchain_api.py]: I'd suggest to not rename the import here.

(and in all following tests)

File: tests/web/test_outputs.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import pytest
+from unittest.mock import MagicMock, patch
+
+pytestmark = [pytest.mark.bdb, pytest.mark.usefixtures('inputs')]
+
+UNSPENTS_ENDPOINT = '/api/v1/outputs/'
```
Comment (TimDaub) [inline tests/web/test_outputs.py]: Rename to `OUTPUTS_ENDPOINT`.

File: tests/db/test_bigchain_api.py [hunk 4]
```diff
@@ -1156,3 +1157,36 @@ def test_get_spent_multiple_owners(self, b, user_sk, user_pk):
...
+        get_outputs.return_value = [TL('a', 1), TL('b', 2)]
+        with patch('bigchaindb.core.Bigchain.get_spent') as get_spent:
+            get_spent.side_effect = [True, False]
+            out = Bigchain().get_outputs_filtered('abc', include_spent=False)
+    get_outputs.assert_called_once_with('abc')
+    assert out == [TL('b', 2)]
+
+
+def test_get_outputs_filtered():
+    from bigchaindb.common.transaction import TransactionLink as TL
```
Comment (TimDaub) [inline tests/db/test_bigchain_api.py]: Same here.

File: bigchaindb/core.py [hunk 5]
```diff
@@ -411,11 +411,30 @@ def get_owned_ids(self, owner):
...
+            owner (str): base58 encoded public key.
+
+        Returns:
+            :obj:`list` of TransactionLink: list of ``txid`` s and ``output`` s
+            pointing to another transaction's condition
+        """
+        return self.get_outputs_filtered(owner, include_spent=False)
 
-        return owned
+    def get_outputs_filtered(self, owner, include_spent=True):
```
Comment (sohkai) [inline bigchaindb/core.py]: I get the impression `get_outputs_filtered` could be merged with `get_outputs`.

File: tests/web/test_outputs.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def test_get_outputs_endpoint(client, user_pk):
+    m = MagicMock()
+    m.to_uri.side_effect = lambda s: s
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
+        gof.return_value = [m, m]
+        res = client.get(UNSPENTS_ENDPOINT + '?public_key={}'.format(user_pk))
+    assert res.json == ["..", ".."]
```
Comment (TimDaub) [inline tests/web/test_outputs.py]: single quotes

File: tests/web/test_outputs.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def test_get_outputs_endpoint_unspent(client, user_pk):
+    m = MagicMock()
+    m.to_uri.side_effect = lambda s: s
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
+        gof.return_value = [m]
+        params = '?unspent=true&public_key={}'.format(user_pk)
+        res = client.get(UNSPENTS_ENDPOINT + params)
+    assert res.json == [".."]
```
Comment (TimDaub) [inline tests/web/test_outputs.py]: single quotes

File: tests/db/test_bigchain_api.py [hunk 8]
```diff
@@ -1156,3 +1157,36 @@ def test_get_spent_multiple_owners(self, b, user_sk, user_pk):
         # check that the other remain marked as unspent
         for unspent in transactions[1:]:
             assert b.get_spent(unspent.id, 0) is None
+
+
+def test_get_owned_ids_calls_get_outputs_filtered():
+    from bigchaindb.core import Bigchain
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
+        b = Bigchain()
+        res = b.get_owned_ids("abc")
```
Comment (TimDaub) [inline tests/db/test_bigchain_api.py]: single quotes

File: tests/db/test_bigchain_api.py [hunk 9]
```diff
@@ -1156,3 +1157,36 @@ def test_get_spent_multiple_owners(self, b, user_sk, user_pk):
...
+    gof.assert_called_once_with("abc", include_spent=False)
```
Comment (TimDaub) [inline tests/db/test_bigchain_api.py]: single quotes

File: tests/web/test_outputs.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        res = client.get(UNSPENTS_ENDPOINT + '?public_key={}'.format(user_pk))
+    assert res.json == ["..", ".."]
+    assert res.status_code == 200
+    gof.assert_called_once_with(user_pk, True)
+
+
+def test_get_outputs_endpoint_unspent(client, user_pk):
+    m = MagicMock()
+    m.to_uri.side_effect = lambda s: s
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
```
Comment (TimDaub) [inline tests/web/test_outputs.py]: Would it make sense to use the patch decorator here?

File: tests/web/test_outputs.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+pytestmark = [pytest.mark.bdb, pytest.mark.usefixtures('inputs')]
+
+UNSPENTS_ENDPOINT = '/api/v1/outputs/'
+
+
+def test_get_outputs_endpoint(client, user_pk):
+    m = MagicMock()
+    m.to_uri.side_effect = lambda s: s
+    with patch('bigchaindb.core.Bigchain.get_outputs_filtered') as gof:
```
Comment (TimDaub) [inline tests/web/test_outputs.py]: Would it make sense to use the patch decorator here?

Also: To my knowledge, the usual `Mock` also lets you define `side_effect`. Why `MagicMock`?

File: tests/common/test_transaction.py [hunk 12]
```diff
@@ -436,6 +436,15 @@ def test_cast_transaction_link_to_boolean():
     assert bool(TransactionLink(False, False)) is True
 
 
+def test_transaction_link_eq():
```
Comment (TimDaub) [inline tests/common/test_transaction.py]: :+1:

File: tests/web/test_outputs.py (hunk 11 above)
Comment (libscott) [inline tests/web/test_outputs.py]: because it's `m.to_uri.side_effect` not `m.side_effect`.

File: tests/web/test_outputs.py (hunk 10 above)
Comment (libscott) [inline tests/web/test_outputs.py]: Is my usage of `patch` somehow wrong?

Comment (TimDaub) [inline tests/web/test_outputs.py]: Sorry. I should have been more explicit here.

As far as I can tell you're using it absolutely correct, but you could save a few lines by using the patch operator. Totally up to you if you wanna go for that though.

File: tests/web/test_outputs.py (hunk 7 above)
Comment (libscott) [inline tests/web/test_outputs.py]: I'm deferring the single quotes issues to #1083

File: bigchaindb/core.py (hunk 5 above)
Comment (libscott) [inline bigchaindb/core.py]: There's a school of thought that suggests that functions should do one thing. This was the idea behind separating the filtering from the fetching.

**Pull request merged**